### PR TITLE
[Updated PR#1534]  UTCDateTime conversion now includes milliseconds

### PR DESCRIPTION
**Pull request by Giacomo92**
This PR was recreated from #1534 
--

Currently laravel-mongodb strips out the millisecond value of the datetime when retrieving records or querying mongo. This PR fixes the date conversion to include the millisecond part of the date. The PR also saves created_at and updated_at with millisecond precision now.

This PR only works for PHP7 and above as I couldn't find anywhere specifying that PHP5 is supported, and travis is only testing on PHP7.

Co-Authored-By: @Flambe

### CODE REVIEW

File: src/Jenssegers/Mongodb/Auth/DatabaseTokenRepository.php [hunk 1]
```diff
@@ -17,7 +17,7 @@ protected function getPayload($email, $token)
         return [
             'email' => $email,
             'token' => $this->hasher->make($token),
-            'created_at' => new UTCDateTime(time() * 1000),
+            'created_at' => new UTCDateTime(now()->format('Uv')),
```
Comment (divine) [inline src/Jenssegers/Mongodb/Auth/DatabaseTokenRepository.php]:
```suggestion
            'created_at' => new UTCDateTime(Date::now()->format('Uv')),
```

File: src/Jenssegers/Mongodb/Eloquent/Model.php [hunk 2]
```diff
@@ -99,7 +99,7 @@ protected function asDateTime($value)
     {
         // Convert UTCDateTime instances.
         if ($value instanceof UTCDateTime) {
-            return Carbon::createFromTimestamp($value->toDateTime()->getTimestamp());
+            return Carbon::createFromTimestampMs($value->toDateTime()->format('Uv'));
```
Comment (divine) [inline src/Jenssegers/Mongodb/Eloquent/Model.php]:
```suggestion
            return Date::createFromTimestampMs($value->toDateTime()->format('Uv'));
```

File: tests/QueryBuilderTest.php [hunk 3]
```diff
@@ -1,6 +1,7 @@
 <?php
 declare(strict_types=1);
 
+use Carbon\Carbon;
```
Comment (divine) [inline tests/QueryBuilderTest.php]:
```suggestion
use Illuminate\Support\Facades\Date;
```

File: tests/QueryBuilderTest.php [hunk 4]
```diff
@@ -545,14 +546,14 @@ public function testUpdateSubdocument()
     public function testDates()
     {
         DB::collection('users')->insert([
-            ['name' => 'John Doe', 'birthday' => new UTCDateTime(1000 * strtotime("1980-01-01 00:00:00"))],
-            ['name' => 'Jane Doe', 'birthday' => new UTCDateTime(1000 * strtotime("1981-01-01 00:00:00"))],
-            ['name' => 'Robert Roe', 'birthday' => new UTCDateTime(1000 * strtotime("1982-01-01 00:00:00"))],
-            ['name' => 'Mark Moe', 'birthday' => new UTCDateTime(1000 * strtotime("1983-01-01 00:00:00"))],
+            ['name' => 'John Doe', 'birthday' => new UTCDateTime(Carbon::parse("1980-01-01 00:00:00")->format('Uv'))],
```
Comment (divine) [inline tests/QueryBuilderTest.php]:
```suggestion
            ['name' => 'John Doe', 'birthday' => new UTCDateTime(Date::parse("1980-01-01 00:00:00")->format('Uv'))],
```

File: src/Jenssegers/Mongodb/Eloquent/Model.php [hunk 5]
```diff
@@ -118,7 +118,7 @@ public function getDateFormat()
      */
     public function freshTimestamp()
     {
-        return new UTCDateTime(Carbon::now());
+        return new UTCDateTime(now()->format('Uv'));
```
Comment (divine) [inline src/Jenssegers/Mongodb/Eloquent/Model.php]:
```suggestion
        return new UTCDateTime(Date::now()->format('Uv'));
```

File: tests/QueryBuilderTest.php (hunk 4 above)
Comment (divine) [reply tests/QueryBuilderTest.php]: and others too, please.

File: src/Jenssegers/Mongodb/Eloquent/Model.php (hunk 2 above)
Comment (Smolevich) [reply src/Jenssegers/Mongodb/Eloquent/Model.php]: What is the profit to use `Date` instead of `Carbon`?

Comment (Smolevich) [reply src/Jenssegers/Mongodb/Eloquent/Model.php]: Sorry, i see this comment https://github.com/jenssegers/laravel-mongodb/pull/1963#pullrequestreview-359367649

**Comment by Smolevich** [comment]
@Giacomo92 resolve conflicts please

**Comment by Flambe** [comment]
@Giacomo92 Thank you for getting this up to date!

**Pull request closed without merging**